dot/core/sync.go

```diff
@@ -50,94 +61,261 @@ func NewSyncer(cfg *SyncerConfig) (*Syncer, error) {
+			if err.Error() == "cannot find parent block in blocktree" {
+				// set request start
+				s.requestStart = s.requestStart - maxResponseSize
+				if s.requestStart < 0 {
+					s.requestStart = 1
+				}
+				log.Debug("[sync] Retrying block request", "start", s.requestStart)
+				go s.sendBlockRequest()
+			} else {
+				log.Error("[sync]", "error", err)
 			}
 
-			s.synced = true
-			return
+		} else {
+			// TODO: max retries before unlocking, in case no response is received
+
+			bestNum, err := s.blockState.BestBlockNumber()
+			if err != nil {
+				log.Crit("[sync] Failed to get best block number", "error", err)
+			} else {
+
+				// check if we are synced or not
+				if bestNum.Cmp(s.highestSeenBlock) >= 0 && bestNum.Cmp(big.NewInt(0)) != 0 {
+					log.Debug("[sync] All synced up!", "number", bestNum)
+
+					if !s.synced {
+						s.lock.Unlock()
+						s.synced = true
+					}
+				} else {
+					// not yet synced, send another block request for the following blocks
+					s.requestStart = highestInResp + 1
+					go s.sendBlockRequest()
+				}
+
+			}
 		}
 
-		time.Sleep(time.Second)
 	}
 }
 
-func (s *Syncer) sendBlockRequest() error {
-	bestNum, err := s.blockState.BestBlockNumber()
-	if err != nil {
-		log.Error("[sync] Failed to get best block number", "error", err)
-		return err
+func (s *Syncer) safeMsgSend(msg network.Message) error {
+	s.chanLock.Lock()
+	defer s.chanLock.Unlock()
+	if s.stopped {
+		return errors.New("service has been stopped")
 	}
+	s.msgOut <- msg
+	return nil
+}
 
+func (s *Syncer) sendBlockRequest() {
 	//generate random ID
 	s1 := rand.NewSource(uint64(time.Now().UnixNano()))
 	seed := rand.New(s1).Uint64()
 	randomID := mrand.New(mrand.NewSource(int64(seed))).Uint64()
 
-	// TODO: can't request from /our/ best block number, need to start requesting from the best block num we have of /theirs/
-	// otherwise there's a chance we might build a block, then miss a block of theirs, causing error="cannot find parent block in blocktree"
-	buf := make([]byte, 8)
-	binary.LittleEndian.PutUint64(buf, uint64(bestNum.Int64()))
+	start, err := variadic.NewUint64OrHash(uint64(s.requestStart))
+	if err != nil {
+		log.Error("[sync] Failed to create StartingBlock", "error", err)
+		return
+	}
+
+	log.Debug("[sync] Block request", "start", start)
 
 	blockRequest := &network.BlockRequestMessage{
 		ID:            randomID, // random
 		RequestedData: 3,        // block header + body
-		StartingBlock: variadic.NewUint64OrHash(append([]byte{1}, buf...)),
+		StartingBlock: start,
 		EndBlockHash:  optional.NewHash(false, common.Hash{}),
 		Direction:     1,
 		Max:           optional.NewUint32(false, 0),
 	}
 
 	// send block request message to network service
-	s.msgOut <- blockRequest
+	err = s.safeMsgSend(blockRequest)
+	if err != nil {
+		log.Error("[sync] Failed to send block request", "error", err)
+	}
+}
 
-	return nil
+func (s *Syncer) handleBlockResponse(msg *network.BlockResponseMessage) (int64, error) {
```

Comment:
This will conflict with #707

Reply:
yes, the contents of this function are nearly the same as `core.ProcessBlockResponse` from before